pkg/distributor/query_ingest_storage_test.go

```diff
@@ -0,0 +1,513 @@
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+					makeWriteRequest(0, 1, 0, false, false, "foo3"),
+					makeWriteRequest(0, 1, 0, false, false, "foo4"),
+				},
+			},
+			preferZone:               "zone-a",
+			minimizeIngesterRequests: true,
+			matchers:                 []*labels.Matcher{selectAllSeriesMatcher},
+			expectedResponse:         expectedResponse(0, 1, false, "foo0", "foo1", "foo2", "foo3", "foo4"),
+			expectedQueriedIngesters: 5 /* zone-a */ + 2, /* the two failed fall back to zone-b */
+		},
+		"should not fallback to ingesters in the non-preferred zone if all ingesters in the preferred zone are healthy and some ingesters in the non-preferred zone are unhealthy": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {numIngesters: 5, happyIngesters: 5},
+				"zone-b": {numIngesters: 5, happyIngesters: 3},
+			},
+			ingesterDataByZone: map[string][]*mimirpb.WriteRequest{
+				"zone-a": {
+					makeWriteRequest(0, 1, 0, false, false, "foo0"),
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+					makeWriteRequest(0, 1, 0, false, false, "foo3"),
+					makeWriteRequest(0, 1, 0, false, false, "foo4"),
+				},
+				"zone-b": {
+					makeWriteRequest(0, 1, 0, false, false, "foo0"),
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+				},
+			},
+			preferZone:               "zone-a",
+			minimizeIngesterRequests: true,
+			matchers:                 []*labels.Matcher{selectAllSeriesMatcher},
+			expectedResponse:         expectedResponse(0, 1, false, "foo0", "foo1", "foo2", "foo3", "foo4"),
+			expectedQueriedIngesters: 5, /* zone-a. zone-b isn't queried because of minimization */
+		},
+		"should fail if all ingesters owning a given partition are unhealthy": {
+			ingesterStateByZone: map[string]ingesterZoneState{
+				"zone-a": {numIngesters: 5, happyIngesters: 3},
+				"zone-b": {numIngesters: 5, happyIngesters: 3},
+			},
+			ingesterDataByZone: map[string][]*mimirpb.WriteRequest{
+				"zone-a": {
+					makeWriteRequest(0, 1, 0, false, false, "foo0"),
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+				},
+				"zone-b": {
+					makeWriteRequest(0, 1, 0, false, false, "foo0"),
+					makeWriteRequest(0, 1, 0, false, false, "foo1"),
+					makeWriteRequest(0, 1, 0, false, false, "foo2"),
+				},
+			},
+			preferZone:               "zone-a",
+			minimizeIngesterRequests: true,
+			matchers:                 []*labels.Matcher{selectAllSeriesMatcher},
+			expectedResponse:         expectedResponse(0, 0, false),
+			expectedQueriedIngesters: 5 /* zone-a */ + 2, /* the two failed fall back to zone-b */
+			expectedErr:              errFail,
+		},
+		"should succeed if there unhealthy ingesters in every zone but unhealthy ingesters own different partitions": {
```

Comment:
This is very nice property.

Reply:
The coolest property of the new system ;)